File: examples/pub_names.rs
//! Calculate the length of the road network (everything tagged `highway=*`)
//! from the input archive.
//!
//! Demonstrates
//!
//!  * iteration through ways
//!  * accessing of tags belonging to a way
//!  * accessing of nodes belonging to a way
//!  * length calculation on the Earth using the haversine function
//!
//! LICENSE
//!
//! The code in this example file is released into the Public Domain.

use osmflat::{Archive, FileResourceStorage, Osm};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let archive_dir = std::env::args()
        .nth(1)
        .ok_or("USAGE: pub_names <osmflat-archive>")?;
    let archive = Osm::open(FileResourceStorage::new(archive_dir))?;

    let nodes_tags = archive.nodes().iter().map(|node| node.tags());
    let ways_tags = archive.ways().iter().map(|way| way.tags());

    let tags = archive.tags();
    let tag_index = archive.tags_index();
    let get_tag = |idx| tags.at(tag_index.at(idx as usize).value() as usize);

    let strings = archive.stringtable();

    for tag_range in nodes_tags.chain(ways_tags) {
        let is_pub = tag_range.clone().map(get_tag).any(|tag| {
            strings[tag.key_idx() as usize..].starts_with(b"amenity\0")
                && strings[tag.value_idx() as usize..].starts_with(b"pub\0")
        });

        if is_pub {
            let name = osmflat::get_tag(&archive, tag_range.clone(), b"name");
            println!(
                "{}",
                name.unwrap_or(Some("broken pub name"))
                    .unwrap_or("unknown pub name")
            );

            let addrs = tag_range
                .clone()
                .map(get_tag)
                .filter(|tag| strings[tag.key_idx() as usize..].starts_with(b"addr:"));
            for tag in addrs {
                let key = strings.substring(tag.key_idx() as usize);
                let value = strings.substring(tag.value_idx() as usize);
                match (key, value) {
                    (Ok(addr_type), Ok(addr)) => println!("  {}: {}", addr_type, addr),
                    _ => (),
                }
            }
        }
    }

    Ok(())
}

File: examples/cpp/pub_names.cpp
#include "osm_generated.h"
#include <cstring>
#include <iostream>

void print_pubs(const osm::Osm &archive, std::pair<uint64_t, uint64_t> range) {
  auto tags = archive.tags();
  auto tags_index = archive.tags_index();
  const char *strings = archive.stringtable().char_ptr();

  bool is_pub = false;
  for (uint32_t idx = range.first; idx < range.second; ++idx) {
    auto tag = tags[tags_index[idx].value];
    const char *key = &strings[tag.key_idx];
    const char *value = &strings[tag.value_idx];
    if (std::strncmp(key, "amenity\0", 8) == 0 && std::strncmp(value, "pub\0", 4) == 0) {
      is_pub = true;
      break;
    }
  }

  if (is_pub) {
    bool has_name = false;
    for (uint32_t idx = range.first; idx < range.second; ++idx) {
      auto tag = tags[tags_index[idx].value];
      const char *key = &strings[tag.key_idx];
      if (std::strncmp(key, "name\0", 5) == 0) {
          const char *value = &strings[tag.value_idx];
          std::cout << value << std::endl;
          has_name = true;
        }
    }
    if (!has_name) {
      std::cout << "unknown pub name" << std::endl;
    }

    for (uint32_t idx = range.first; idx < range.second; ++idx) {
      auto tag = tags[tags_index[idx].value];
      const char *key = &strings[tag.key_idx];
      if (std::strncmp(&strings[tag.key_idx], "addr:", 5) == 0) {
        const char *value = &strings[tag.value_idx];
        std::cout << "  " << key << ": " << value << std::endl;
      }
    }
  }
}

int main(int argc, char const *argv[]) {
  if (argc != 2) {
    std::cerr << "USAGE: pub_name <osmflat-archive>" << std::endl;
    return 1;
  }

  auto storage = flatdata::FileResourceStorage::create(argv[1]);
  auto archive = osm::Osm::open(std::move(storage));

  for (auto node : archive.nodes()) {
    print_pubs(archive, node.tags);
  }

  for (auto way : archive.ways()) {
    print_pubs(archive, way.tags);
  }

  return 0;
}
